File: lang-py/No503.py
"""
  author <a href="https://github.com/Matthew-Han">Matthew Han</a>
  since 2024/7/9 下午3:21:24
"""
from collections import defaultdict, OrderedDict
from typing import List


class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        return list(self.find_idx(nums, idx) for idx in range(len(nums)))

    def find_idx(self, nums: List[int], idx: int) -> int:
        i = (idx + 1) % len(nums)
        while i < len(nums):
            if i == idx:
                return -1
            if nums[i] > nums[idx]:
                return nums[i]
            i = (i + 1) % len(nums)

File: lang-js/No503.js
/**
 * @author <a href="https://github.com/Matthew-Han">Matthew Han</a>
 * @since 2024/7/9 下午4:21:58
 **/
let nextGreaterElements = (nums) => {
    let ans = []
    for (let i = 0; i < nums.length; i++) {
        ans[i] = findIdx(nums, i)
    }
    return ans
};

let findIdx = (nums, idx) => {
    let i = (idx + 1) % nums.length
    while (i < nums.length) {
        if (i === idx) {
            return -1
        }
        if (nums[i] > nums[idx]) {
            return nums[i]
        }
        i = (i + 1) % nums.length
    }
    return -1
}